internal/worker/sshserver/machine/sftp.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package machine
+
+import (
+	"io"
+	"sync"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/juju/errors"
+	gossh "golang.org/x/crypto/ssh"
+)
+
+// SFTPHandler returns a handler for the SFTP subsystem.
+// It requests an sftp subsystem on the target machine
+// and proxies the user's session to it.
+func (s *Handlers) SFTPHandler() ssh.SubsystemHandler {
+	return func(session ssh.Session) {
+		handleError := func(err error) {
+			s.logger.Errorf("sftp proxy failure: %v", err)
+			_, _ = session.Stderr().Write([]byte(err.Error() + "\n"))
+			_ = session.Exit(1)
+		}
+
+		client, err := s.connector.Connect(s.destination)
+		if err != nil {
+			handleError(errors.Annotate(err, "failed to connect to machine"))
+			return
+		}
+		defer client.Close()
+
+		machineChannel, machineReqs, err := client.OpenChannel("session", nil)
+		if err != nil {
+			handleError(errors.Annotate(err, "failed to request session"))
+			return
+		}
+		defer machineChannel.Close()
+		go proxyReqs(session, machineReqs, s.logger)
+
+		err = requestSubsystem(machineChannel, "sftp")
+		if err != nil {
+			handleError(errors.Annotate(err, "failed to request subsystem"))
+			return
+		}
+
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+
+		// Note that we don't close the session immediately after copying
+		// in order to propagate the exit code back to the client.
+
+		go func() {
+			defer wg.Done()
+			defer machineChannel.Close()
+			_, _ = io.Copy(machineChannel, session)
+		}()
+
+		go func() {
+			defer wg.Done()
+			defer machineChannel.Close()
+			_, _ = io.Copy(session, machineChannel)
```

Comment:
So when the sftp subsystem closes, is the machineChannel (or underlying connection) closed? Thus stopping these routines?

Reply:
When one side closes closes the connection, the io.Copy() receives an EOF and that will close the machine channel if it's not already closed.
Inside `proxyReqs`, when the machineChannel is closed, it will close the client' session.